Build System: 🏗️ vcpkg integration for ITGmania build system

sukibaby opened the pull request:
I have been using this vcpkg integration in my dev branches for the last couple months. It works spectacularly and significantly reduces headache as well as build time!

I have tested this very thoroughly on Windows, macOS and Debian. With this change, updating dependencies is as simple as updating `vcpkg.json`.

# What this PR does

The main changes:

 - Allows the possibility to get **all dependencies** except lua and Simply Love via vcpkg, offering the possibility to reduce # of submodules
 - Leaves the traditional non-vcpkg build completely intact
 - eliminates the need to manually maintain CMake for building libraries
 - simplies upgrading of libraries and/or pinning of versions

Smaller, but still notable changes:
 - Provides clear, concise, multi-platform build instructions in `Build/README.md`
 - Updates CryptManager so that the custom libtommath behavior we depend on is available without the custom build
 - Updates the OpenGL driver to remove an unneeded call that is not present in vcpkg's `glew`

# Building ITGmania using vcpkg

<details> 

### 1. Clone and Bootstrap vcpkg

```bash
git clone https://github.com/microsoft/vcpkg.git ~/vcpkg
cd ~/vcpkg
./bootstrap-vcpkg.sh (this is .bat on windows)
```

### 2. Set Environment Variable

Add vcpkg to your shell configuration:

**macOS/Linux (add to `~/.zshrc` or `~/.bashrc`):**
```bash
export vcpkg_ROOT=$HOME/vcpkg
```

**Windows (PowerShell):**
```powershell
$env:vcpkg_ROOT = "C:\path\to\vcpkg"
```

### 3. Install Dependencies

Navigate to the ITGmania source directory and install vcpkg packages:

**macOS (arm64):**
```bash
cd /path/to/itgmania
$vcpkg_ROOT/vcpkg install --triplet arm64-osx
```

**macOS (x86_64):**
```bash
$vcpkg_ROOT/vcpkg install --triplet x64-osx
```

**Windows:**
```bash
cd \path\to\itgmania
$env:vcpkg_ROOT\vcpkg install --triplet x64-windows-static
```

**Linux:**
```bash
cd /path/to/itgmania
$vcpkg_ROOT/vcpkg install --triplet x64-linux
```

### 4. Configure with CMake

```bash
cmake -B Build -DSM_USE_VCPKG=ON -DCMAKE_TOOLCHAIN_FILE=$vcpkg_ROOT/scripts/buildsystems/vcpkg.cmake
```

### 5. Build

**macOS/Linux:**
```bash
cmake --build Build -j $(nproc)
```

**Windows:**
```bash
cmake --build Build -j %NUMBER_OF_PROCESSORS%
```

</details>


# Building ITGmania without vcpkg

<details>

```bash
cmake -B Build -DSM_USE_VCPKG=OFF
cmake --build Build
```

</details>

sukibaby commented:
Even though this works locally, it fails because CI is blocked from downloading from sourceforge. As a result, CI can't download `pcre`, so it fails.

Ironically, this PR is intended to improve build/CI reliability, but it's not doing a good job of gaining anyone's confidence in that regard lol.

I reported the issue to vcpkg. Feel free to clone it locally and try it out in the meantime. I will push an update to re-run CI when it is resolved.

sukibaby committed:
Switch to vcpkg for dependency management

Migrates external library management to vcpkg, which is inherently optional and does not affect the ability to build without vcpkg. However, with this change, the majority of submodules would become optional.

ScottBrenner commented on vcpkg.json:
How about adding vcpkg to the Dependabot configuration?: https://github.com/sukibaby/itgmania-testing/blob/vcpkg-pr/.github/dependabot.yml

They added support recently - 

- https://github.blog/changelog/2025-08-12-dependabot-version-updates-now-support-vcpkg/

- https://docs.github.com/en/code-security/dependabot/ecosystems-supported-by-dependabot/supported-ecosystems-and-repositories?versionId=free-pro-team%40latest&productId=code-security&restPage=dependabot%2Cworking-with-dependabot%2Cdependabot-options-reference#vcpkg

sukibaby replied:
Thank you for the suggestion! There are some nice benefits to having dependabot support vcpkg, so I've added support to `dependabot.yml`.  I think I did it correctly, but I'm always nervous working with yaml, so let me know if it's not correct

ScottBrenner commented on .github/workflows/ci.yml:
```diff
@@ -16,54 +16,64 @@ jobs:
+          if [ ! -d "$HOME/vcpkg" ]; then
+            git clone https://github.com/microsoft/vcpkg.git $HOME/vcpkg
+            cd $HOME/vcpkg
+            ./bootstrap-vcpkg.sh
```
```suggestion
            ./bootstrap-vcpkg.sh -disableMetrics
```

ScottBrenner commented on .github/workflows/codeql.yml:
```diff
@@ -55,8 +59,19 @@ jobs:
+        if [ ! -d "$HOME/vcpkg" ]; then
+          git clone https://github.com/microsoft/vcpkg.git $HOME/vcpkg
+          cd $HOME/vcpkg
+          ./bootstrap-vcpkg.sh
```
```suggestion
          ./bootstrap-vcpkg.sh -disableMetrics
```

ScottBrenner reviewed (commented):
https://github.com/itgmania/itgmania/actions/runs/19256909238/job/55053166706#step:4:18

sukibaby committed:
Disable vcpkg telemetry in codeql.yml

Co-authored-by: Scott Brenner <scott@scottbrenner.me>

dinsfire64 commented:
Fails to build using traditional cmake build system on current Arch on ixwebsocket.

`Linux hyrule 6.17.7-arch1-1 #1 SMP PREEMPT_DYNAMIC Sun, 02 Nov 2025 17:27:22 +0000 x86_64 GNU/Linux`

I have not tried vcpkg as I am interested in maintaining compatibility with existing systems.

```
In file included from /home/dinsfire/Downloads/test/itgmania-testing/src/StepMania.cpp:64:
/home/dinsfire/Downloads/test/itgmania-testing/src/NetworkManager.h:14:10: fatal error: ixwebsocket/IXHttp.h: No such file or directory
   14 | #include <ixwebsocket/IXHttp.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

itgmania's beta branch builds correctly, attached is cmake's output and beta branch's cmake output.

[cmake.log](https://github.com/user-attachments/files/23479876/cmake.log)
[cmake-betabranch.log](https://github.com/user-attachments/files/23479893/cmake-betabranch.log)

NOTE: reverting the `extern/CMakeLists.txt` to the previous version yields a successful compile.

Maxwell-lt commented:
I'm seeing the same error as @dinsfire64 when building this PR branch with Nix, where IXHttp.h and json.h are not found. Building from itgmania:beta (506126d76cc748167320ce31a3092c88d730256c) works fine.

[nix-vcpkg.log](https://github.com/user-attachments/files/23483842/nix-vcpkg.log)

ScottBrenner commented:
At a glance vcpkg appears to be included in GitHub's runner images - https://github.com/search?q=repo%3Aactions%2Frunner-images+path%3A%2F%5Eimages%5C%2F%2F+vcpkg+language%3AMarkdown&type=code&l=Markdown - and would not need an additional installation step here

Similarly, the checks have a few annotations about other packages that are included in the runner image and do not need to be included in the installation step as well - https://github.com/itgmania/itgmania/actions/runs/19257524914

aryla commented:
I tried this in an Ubuntu 22.04 Podman container. Linking failed with this error:
```
[ 59%] Linking CXX executable ../../itgmania-release-symbols
/usr/bin/ld: ../ffmpeg-prefix/src/ffmpeg-build/dest/lib/libavcodec.a(exrenc.o): undefined reference to symbol 'compress'
/usr/bin/ld: /lib/x86_64-linux-gnu/libz.so.1: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
gmake[2]: *** [src/CMakeFiles/ITGmania.dir/build.make:7011: ../itgmania-release-symbols] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:286: src/CMakeFiles/ITGmania.dir/all] Error 2
gmake: *** [Makefile:156: all] Error 2
```
I guess libavcoded depends on zlib, so zlib needs to be in the library list after it. Making this change to `src/CMakeLists.txt` fixed it for me:
```diff
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -481,6 +481,9 @@ else() # Unix / Linux TODO: Remember to find and locate the zip archive files.
   if(NOT (DEFINED CMAKE_TOOLCHAIN_FILE AND CMAKE_TOOLCHAIN_FILE MATCHES "vcpkg.cmake"))
     list(REMOVE_ITEM SMDATA_LINK_LIB "zlib")
     list(APPEND SMDATA_LINK_LIB "zlib")
+  else()
+    list(REMOVE_ITEM SMDATA_LINK_LIB ZLIB::ZLIB)
+    list(APPEND SMDATA_LINK_LIB ZLIB::ZLIB)
   endif()

   list(APPEND SMDATA_LINK_LIB ${CMAKE_DL_LIBS}
```

When running cmake without `-DCMAKE_TOOLCHAIN_FILE`, the build fails the same `json.h` error that @Maxwell-lt reported.

sukibaby committed:
Add configurable vcpkg toolchain support to CMake

Introduces the SM_USE_VCPKG option to control usage of the vcpkg toolchain, with improved error handling and messaging. Updates extern/CMakeLists.txt to conditionally include third-party projects based on whether vcpkg is used, ensuring local forks and custom options are always built.

sukibaby committed:
Add build instructions to CMakeLists.txt

sukibaby commented:
Updated and the traditional build seems to be working as intended again, but please lmk.

The build commands are slightly changed with the new commits:

With vcpkg:

```
cmake -S . -B Build \
  -DSM_USE_VCPKG=ON \
  -DCMAKE_TOOLCHAIN_FILE="$VCPKG_ROOT/scripts/buildsystems/vcpkg.cmake"
```

Without vcpkg:

```
cmake -S . -B Build \
  -DSM_USE_VCPKG=OFF \
  -DCMAKE_TOOLCHAIN_FILE:PATH=
```

Then run  cmake --build build as usual.

Maxwell-lt commented:
Can confirm that the Nix build is working again as of eed9127

sukibaby committed:
Enable SM_USE_VCPKG in CI and CodeQL workflows

Necessary following the previous commit.

sukibaby committed:
Prevent macro redefinition with tomcrypt

Revert "Remake release.yml for vcpkg"

This reverts commit 07d663f7c6c85b8a3d155588bd09f91328b5b051.

Remake release.yml for vcpkg

sukibaby committed:
Remove unnecessary comment

sukibaby committed:
Match vcpkg.json versions to extern submodules

glew: same version in vcpkg.json and submodule (vcpkg version may have additional patches)

hidapi: same version in vcpkg.json and submodule (vcpkg version may have additional patches)

ixwebsocket: 11.4.6 in vcpkg.json, 11.4.5 in extern dir. leaving at 11.4.6 since networking is stable on all OS’s with new mbedtls version (see below). https://github.com/machinezone/IXWebSocket/releases/tag/v11.4.6

libjpeg-turbo: 3.1.2 in vcpkg.json, 3.1.0 in extern. rolling back to 3.1.0#2.

libmad: 0.16.4#4 in vcpkg.json, 0.15 in extern. leaving at 0.16.4#4 since upgrading to the tenacityteam fork was previously discussed but deferred.

libogg: 1.3.6#1 in vcpkg.json, 1.3.5 in extern. rolling back to 1.3.5#2

libpng: 1.6.50 in vcpkg.json, 1.6.48 in submodule. rolling back to 1.6.48#0

libvorbis: same version in vcpkg.json and submodule (vcpkg version may have additional patches)

jsoncpp: 1.9.6 in vcpkg.json, 1.9.5 in submodule. using vcpkg version since it contains bugfixes and security fixes. https://github.com/open-source-parsers/jsoncpp/releases/tag/1.9.6 - recommend updating submodule

mbedtls: 3.6.4 in vcpkg.json, 3.5.2 in submodule. leaving at 3.6.4 because it resolves the windows issues which were why we had to roll back from 3.6.0. would recommend updating submodule.

miniz: 3.1.0 in vcpkg, 2.2.0 in submodule. rolling back to 2.2.0#1.

pcre: same version in vcpkg.json and submodule (vcpkg version may have additional patches)

zlib: same version in vcpkg.json and submodule

sukibaby committed:
Update Build Documentation for vcpkg build

Also necessary to make an exception in the gitignore file for these new Markdown documents.

sukibaby committed:
Update zlib linking logic for vcpkg toolchain

Adjusts SMDATA_LINK_LIB to use ZLIB::ZLIB when building with vcpkg toolchain, ensuring correct linkage on Unix/Linux systems. Fix by @aryla (noticed on Ubuntu 22.04 / Podman).

Co-Authored-By: Arttu Ylä-Outinen <7724847+aryla@users.noreply.github.com>

sukibaby committed:
Update vcpkg baseline and CI workflow setup

Revert "Update vcpkg baseline and CI workflow setup"

This reverts commit cfce4c40c98b31b392fcf9a076bd2e78b0def912.

Update vcpkg builtin-baseline hash

Changed the builtin-baseline in vcpkg.json to a newer commit to ensure dependency versions are up to date.

sukibaby commented:
Doing a bit of maintenance following the rebase. Will mark as ready for review once tested on all three OS's again.

sukibaby committed:
works locally... hope ci passes :3

sukibaby committed:
Restore CI instrctions to Build/README.md

Also rewrites it a bit so it's easier for less technical users to understand.

sukibaby committed:
Correct libtool comment in Build/README.md

ScottBrenner commented on .github/workflows/ci.yml:
```diff
@@ -16,54 +16,77 @@ jobs:
+            cd $HOME/vcpkg
+            ./bootstrap-vcpkg.sh -disableMetrics
+          fi
+          echo "VCPKG_ROOT=$HOME/vcpkg" >> $GITHUB_ENV
```
```suggestion
```
No harm in leaving this I suppose but it is already setup in the runner images - https://github.com/itgmania/itgmania/pull/1065#issuecomment-3519636006 - albeit with a different path/environment variable https://github.com/actions/runner-images/blob/main/images/ubuntu/Ubuntu2404-Readme.md#environment-variables

sukibaby replied:
good point, also gets ci.yml under 100 lines with this change, pretty cool

ScottBrenner commented on .github/workflows/codeql.yml:
```diff
@@ -55,8 +59,20 @@ jobs:
+          cd $HOME/vcpkg
+          ./bootstrap-vcpkg.sh -disableMetrics
+        fi
+        echo "VCPKG_ROOT=$HOME/vcpkg" >> $GITHUB_ENV
```
```suggestion
        echo "VCPKG_ROOT=$VCPKG_INSTALLATION_ROOT" >> $GITHUB_ENV
```

sukibaby committed:
Further consolidation of vcpkg use in CI

Co-authored-by: Scott Brenner <scott@scottbrenner.me>

ScottBrenner commented on .github/dependabot.yml:
```diff
@@ -15,3 +15,9 @@ updates:
+    directory: "/"
+    target-branch: "beta"
+    schedule:
+      interval: "semiannually"
```
```suggestion
      interval: "monthly"
```
Or quarterly? https://docs.github.com/en/code-security/reference/supply-chain-security/dependabot-options-reference#schedule-

sukibaby replied:
My reason for choosing `semiannually` is to align with ITGmania's semiannual release cycle. I am not too picky about this though;  I mainly thought this would minimize PR noise by only doing the version check 2x a year, and also this would `align well with our release cycle, but if teejusb wants it more frequent than that then no problem there.

ScottBrenner replied:
Makes sense, figure getting things into the beta branch more frequently might allow for more testing before the releases

sukibaby replied:
On further contemplation I've seen that other dependabots in the repo have a `monthly` timing, so I'll update to match

sukibaby committed:
Update .github/dependabot.yml

Co-Authored-By: Scott Brenner <scott@scottbrenner.me>

sukibaby committed:
Remove GLU_VERSION check

It would be literally impossible to run this game on a system that doesn't have OpenGL 1.2 types. Additionally the GLU_VERSION call isn't recognized by vcpkg's glew.

sukibaby committed:
vcpkg yaml update

Remake release.yml for vcpkg

Use vcpkg in ci.yml also

Update ci.yml

Update release.yml

Add back autoconf and libtool to macOS workflows

Needed to build libusb via vcpkg.

resolve merge conflicts with recent YAML edits

Add back autoconf and libtool to macOS workflows

Needed to build libusb via vcpkg.

Update release.yml

Update ci.yml

Use vcpkg in ci.yml also

Remake release.yml for vcpkg

sukibaby committed:
mac does want automake actually

sukibaby committed:
and also re-add automake to ci.yml
